Let aotriton.cmake detect the best binary package to use, and deprecate aotriton_version.txt

xinyazhang opened the pull request:
We do not need `install_aotriton.sh` and `aotriton_version.txt` any more since `aotriton.cmake` now installs the best binary release package as the default option when building pytorch.

This should resolve the issue of needing a pre-installed aotriton package when building PyTorch for ROCm from source, which is not feasible if building PyTorch *outside* a CI docker image. With this change, a user can have a pre-installed AOTriton in their environment, if desired, and have the build pick it up by specifying the `AOTRITON_INSTALLED_PREFIX` env var, or have the build automatically detect and install the compatible version. As a third option, the user can also force AOTriton to build from source instead, using the `AOTRITON_INSTALL_FROM_SOURCE` env var.

Also, with the changes in this PR, the cmake build process handles the tasks of copying aotriton .so and images directory from `torch/lib` to the installation path.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

xinyazhang committed:
Merge aotriton_version.txt into aotriton.cmake

xinyazhang committed:
Deprecate install_aotriton.sh

xinyazhang commented:
@jithunnair-amd shall we use this PR to bump to 0.7.1b as well?

jithunnair-amd reviewed (approved):
LGTM, let's make sure various CI workflows run successfully

jithunnair-amd commented:
> @jithunnair-amd shall we use this PR to bump to 0.7.1b as well?

No, I'd prefer we do that in a different PR

xinyazhang committed:
Fix list(join)

Should use string(join)

jithunnair-amd commented on cmake/External/aotriton.cmake:
```diff
@@ -37,17 +56,27 @@ if(NOT __AOTRITON_INCLUDED)
+    elseif(__AOTRITON_SYSTEM_ROCM VERSION_GREATER __AOTRITON_ROCM_HIGH)
+      set(__AOTRITON_ROCM ${__AOTRITON_ROCM_HIGH})
+    else()
+      set(__AOTRITON_ROCM ${__AOTRITON_SYSTEM_ROCM})
```
Does this logic not require publishing aotriton tarballs for each ROCm version, even if there's no ABI breakage from one ROCm version to the next?

xinyazhang replied:
Yes, but it won't happen unless AOTriton decides to skip a certain version of ROCM (which is unlikely)

xinyazhang replied:
Done

jithunnair-amd commented on cmake/External/aotriton.cmake:
```diff
@@ -4,6 +4,27 @@ if(NOT __AOTRITON_INCLUDED)
+  # AOTriton package information from GitHub Release Pages
+  # Replaces .ci/docker/aotriton_version.txt
+  set(__AOTRITON_VER "0.7b")
+  string(JOIN ";"
+         __AOTRITON_MANYLINUX_LIST
+         "manylinux_2_17"  # rocm6.1
+         "manylinux_2_17"  # rocm6.2
+         )
```
```suggestion
set(__AOTRITON_MANYLINUX_LIST
    "manylinux_2_17"  # rocm6.1
    "manylinux_2_17"  # rocm6.2
   )
```

This can be a little less verbose if the intention is to create a CMake list? Likewise for `__AOTRITON_ROCM_LIST` and `__AOTRITON_SHA256_LIST`

xinyazhang replied:
If I understand CMake's list correctly, it is is a string with `;` as separator, and the suggested change will make it a string instead of list

jithunnair-amd replied:
IIUC from https://cmake.org/cmake/help/latest/command/list.html, `set(var a b c d e) creates a list with a;b;c;d;e`, so specifying the ";" is not mandatory. I tested the suggested change locally and it correctly picked the right value from the list.

xinyazhang committed:
New package selection algorithm

Select the last package built with rocm <= system ROCm. This allows
AOTrtion to skip ROCM builds that do not have ABI breaks.

Example:
AOTriton has builds on ROCm 6.1 6.2 6.4

| System ROCm  | Use AOTriton Package |
|--------------------------------------
|  6.0         |       6.1            |
|  6.1         |       6.1            |
|  6.2         |       6.2            |
|  6.3         |       6.2            |
|  6.4         |       6.4            |

xinyazhang commented:
Confirmed the new algorithm's correctness with
``` cmake

set(__AOTRITON_ROCM_LIST
    "rocm6.1"
    "rocm6.2"
    "rocm6.4"
)

set(detection_lists
    "6.0"
    "6.1"
    "6.2"
    "6.3"
    "6.4"
)
foreach(__AOTRITON_SYSTEM_ROCM IN LISTS detection_lists)
    list(GET __AOTRITON_ROCM_LIST 0 __AOTRITON_ROCM_DEFAULT_STR)
    # Initialize __AOTRITON_ROCM to lowest version, in case all builds > system's ROCM
    string(SUBSTRING ${__AOTRITON_ROCM_DEFAULT_STR} 4 -1 __AOTRITON_ROCM)
    foreach(AOTRITON_ROCM_BUILD_STR IN LISTS __AOTRITON_ROCM_LIST)
      # len("rocm") == 4
      string(SUBSTRING ${AOTRITON_ROCM_BUILD_STR} 4 -1 AOTRITON_ROCM_BUILD)
      # Find the last build that <= system's ROCM
      # Assume the list is from lower to higher
      if(AOTRITON_ROCM_BUILD VERSION_GREATER __AOTRITON_SYSTEM_ROCM)
        break()
      endif()
      set(__AOTRITON_ROCM ${AOTRITON_ROCM_BUILD})
    endforeach()
    message(STATUS "System ROCM ${__AOTRITON_SYSTEM_ROCM}, Selected Package ${__AOTRITON_ROCM}")
endforeach()
```
Output
```
-- System ROCM 6.0, Selected Package 6.1
-- System ROCM 6.1, Selected Package 6.1
-- System ROCM 6.2, Selected Package 6.2
-- System ROCM 6.3, Selected Package 6.2
-- System ROCM 6.4, Selected Package 6.4
```

xinyazhang commented:
Current plan: let's send varlen/nestedtensor support, which requires some fixes in AOTriton 0.7.2b, and then update this PR.

huydhn commented:
@pytorchbot rebase

jithunnair-amd commented:
We'll hold on this PR until aotriton 0.8 PR is merged

jithunnair-amd committed:
Update info for aotriton 0.8b

jithunnair-amd committed:
Add Navi31 target for build from source since 0.8b supports it

jithunnair-amd committed:
PyTorch build should take care of copying aotriton lib and images

xinyazhang commented:
@jithunnair-amd 
![image](https://github.com/user-attachments/assets/0c811266-3398-4f96-9e21-4b64a06373fe)
Seems existing logic already copies the files under `torch/lib`, and we can move this out of draft status.

jithunnair-amd commented:
Confirmed that the build log shows copying of the `aotriton.images` and `libaotriton_v2.so` from the `torch/lib` directory: 
```
2024-12-24T05:16:11.3127478Z creating build/lib.linux-x86_64-cpython-39/torch/lib/aotriton.images/amd-gfx110x/flash/attn_fwd
2024-12-24T05:16:11.3128737Z copying torch/lib/aotriton.images/amd-gfx110x/flash/attn_fwd/FONLY__^fp16@16,False,256,False,True,False,1___Navi3x.aks2 -> build/lib.linux-x86_64-cpython-39/torch/lib/aotriton.images/amd-gfx110x/flash/attn_fwd
2024-12-24T05:16:11.3153117Z copying torch/lib/aotriton.images/amd-gfx110x/flash/attn_fwd/FONLY__^fp16@16,True,256,False,False,False,0___Navi3x.aks2 -> build/lib.linux-x86_64-cpython-39/torch/lib/aotriton.images/amd-gfx110x/flash/attn_fwd
2024-12-24T05:16:11.3174282Z copying torch/lib/aotriton.images/amd-gfx110x/flash/attn_fwd/FONLY__^bf16@16,False,128,False,True,False,0___Navi3x.aks2 -> build/lib.linux-x86_64-cpython-39/torch/lib/aotriton.images/amd-gfx110x/flash/attn_fwd
2024-12-24T05:16:11.3192749Z copying torch/lib/aotriton.images/amd-gfx110x/flash/attn_fwd/FONLY__^fp16@16,True,128,True,False,True,0___Navi3x.aks2 -> build/lib.linux-x86_64-cpython-39/torch/lib/aotriton.images/amd-gfx110x/flash/attn_fwd

```

jithunnair-amd commented:
@pytorchbot merge -f "Rocm CI jobs passed. Wheel build jobs also passed, installing all aotriton files during Pytorch build as expected"
